[API PULL] POST method for SYNC endpoint

puntope opened the pull request:
### Changes proposed in this Pull Request:

In this PR we create POST method for the SYNC endpoint `/wp-json/wc/gla/sync` and save it inside `wp_options.gla_api_pull_sync_mode`

PT: pcTzPl-2yG-p2

### Screenshots:

<img width="1278" alt="Screenshot 2025-02-18 at 11 48 32" src="https://github.com/user-attachments/assets/98de74c4-3943-40f7-b4f0-675d0aeb6704" />


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Request POST wp-json/wc/gla/sync with some valid values and some random values (a mix with good values and wrong keys and wrong values) 
2. Verify that only the valid keys and values are updated in the options table. Everything is optional. Invalid or missing datatypes or methods won't be updated.  


Examples

1. Updates all data types
```
'products' => [
			'pull' => bool,
			'push' => bool,
		],
		'coupons'  => [
			'pull' => bool,
			'push' => bool,
		],
		'shipping' => [
			'pull' => bool,
			'push' => bool,
		],
		'settings' => [
			'pull' => bool,
			'push' => bool,
		],
```

2. Updates products data type
```
'products' => [
			'pull' => bool,
			'push' => bool,
		],
]
```

2. Updates products pull method

```
'products' => [
			'pull' => bool
]		],
```


### Additional details:

Didn't add yet `gla_syncable` as a required param because anyone can really add it as a param in the request to bypass the check ...  so it doesn't make much difference to use it as a protection layer. Also even if we protect the API anyone can just update the options table for changing the modes. So for now I just keep the auth access as `can_manage` as we do for other endpoints.


<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

>

puntope committed:
Add post method for Sync Endpoint

puntope committed:
Handle when Post Body is encoded

ianlin commented on src/API/Site/Controllers/RestAPI/SyncController.php:
```diff
@@ -146,4 +166,57 @@ private function get_pull_push_schema_fields() {
+		// JSON Data
+		if ( is_null( $request_params ) ) {
+			$request_params = json_decode( $request->get_body(), true );
+		}
```
From my local test the `$request->get_json_params()` will be `null` if the content type is not `application/json`. For `$request->get_body()`  it accepts multiple content types. In this case not sure if we can just always use `get_body()`?

Also from the local test if I don't provide body and the 500 will return.

puntope replied:
Perfect @ianlin  I removed the get_json_params part.

ianlin reviewed (commented):
Thanks for adding the POST sync endpoint in the plugin. The behaviour works as expected in my local. I left a question regarding request body parsing.

mikkamp commented:
I had another question to clarify about this PR.

Taking a quick look at the code it seems that I can update a single group at a time. Which will cause the updated group to be merged with the current sync modes. However I'm not seeing that clarified in the testing instructions or reflected in the schema. Can we document a little more on how this endpoint works, either in the schema so API self discovery can be used, or somewhere else to be clear what is expected.

Same goes for the possibility to set an individual sync mode, for example if I want to set:

```json
{
	"products": {
		"pull": true
	}
}
```

From my reading of the code, that's not possible as both the pull and push should be set at the same time. I'm not saying we should necessarily support this, but we should evaluate whether that's needed. If not then it should be made clear that it's not supported.

puntope committed:
Refactor Sync controller

puntope commented:
@ianlin @mikkamp Thanks for the review and comments. I addressed them by adding:

- Updates what is sent in the body. The rest will remain as it was. This applies to Data Type (Products, Coupon...) and methods (Push or Pull). Notice that before adding both push and pull were required. 
- Add docs at the top of the class with examples.
- Methods to prevent malformations were added to the option. For example. Suppose someone saves in the option a string or any other thing outside the allowed structure, then the default value would be used. 
- Remove unnecessary get_json_params

puntope commented:
> I had another question to clarify about this PR.
> 
> Taking a quick look at the code it seems that I can update a single group at a time. Which will cause the updated group > to be merged with the current sync modes. 

Correct. 

> Same goes for the possibility to set an individual sync mode. 
> From my reading of the code, that's not possible as both the pull and push should be set at the same time. 

Correct as well. Now I updated it and is allowed.

> However I'm not seeing that clarified in the testing instructions or reflected in the schema

[The Schema was added here](https://github.com/woocommerce/google-listings-and-ads/pull/2796/files). Is there something missing? I added a description. Not sure if something else is expected. 

About PR description. Sorry, fair enough. 

The PR description says: 2. Verify that only the valid keys and values are updated in the options table. 
With "valid keys" I meant the ones set in the body. But you're right its a bit vague. I will update it and also I will add docs in the controller.

ianlin reviewed (approved):
Thanks for the changes. LGTM

- Confirmed the 500 error is no longer showing when the body of the POST request is empty.
- Can update an individual sync mode.
- The doc explains clearly
